File: src/Environment.ts
import type { LiteralType } from './type';
import type Token from './Token';
class Environment {
  private readonly values = new Map<string, LiteralType>();
  private parent: Environment | null = null;
  constructor(parent: Environment | null) {
    this.parent = parent;
  }
  get(name: Token): LiteralType {
    if (this.values.has(name.lexeme)) {
      return this.values.get(name.lexeme) as LiteralType;
    }
    if (this.parent !== null) {
      return this.parent.get(name);
    }
    throw new Error(`${name.lexeme} is not defined`);
  }
  define(name: string, value: LiteralType) {
    this.values.set(name, value);
  }
  assign(name: Token, value: LiteralType) {
    if (this.values.has(name.lexeme)) {
      this.values.set(name.lexeme, value);
      return;
    }
    if (this.parent !== null) {
      this.parent.assign(name, value);
      return;
    }
    throw new Error(`${name.lexeme} is not defined`);
  }
}

export default Environment;

File: environment.go
package main

import (
	"fmt"
)

type Environment struct {
	parent *Environment
	values map[string]any
}

func NewEnvironment(parent *Environment) *Environment {
	values := make(map[string]any)
	return &Environment{
		parent: parent,
		values: values,
	}
}

func (environment *Environment) get(name *Token) any {
	if val, ok := environment.values[name.lexeme]; ok {
		return val
	}
	if environment.parent != nil {
		return environment.parent.get(name)
	}
	panic(any(fmt.Sprintf("%s is not defined", name.lexeme)))
}
func (environment *Environment) define(name string, value any) {
	environment.values[name] = value
}

func (environment *Environment) assign(name *Token, value any) {
	if _, ok := environment.values[name.lexeme]; ok {
		environment.define(name.lexeme, value)
		return
	}
	if environment.parent != nil {
		environment.parent.assign(name, value)
		return
	}
	panic(any(fmt.Sprintf("%s is not defined", name.lexeme)))
}
